### PR TITLE
Make all the piece processors picklable

### PR DESCRIPTION
**Pull request by danieldk**
## Description

The piece processors are not pure Python classes, so could not be pickled. Implement `__reduce__` on all the processors to make them pickable.

This is one step in making spaCy multiprocessing work.

See: https://github.com/explosion/spaCy/issues/13233

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_tokenizers/_spp.pyx [hunk 1]
```diff
@@ -230,3 +233,6 @@ cdef _check_status(Status status):
         raise RuntimeError(status.error_message().decode("utf-8"))
     else:
         raise ValueError(status.error_message().decode("utf-8"))
+
+def unpickle_sentence_piece_processor(protobuf: bytes):
```
Comment (svlandeg) [inline curated_tokenizers/_spp.pyx:237]:
```suggestion
def unpickle_sentence_piece_processor(protobuf: bytes) -> SentencePieceProcessor:
```
?

File: curated_tokenizers/tests/test_bbpe_processor.py [hunk 2]
```diff
@@ -114,3 +115,10 @@ def test_can_load_from_file_object(vocab_path, merges_path):
                 merges=merges,
             )
     assert toy_processor.encode(EXAMPLE_TEXT) == (EXAMPLE_PIECE_IDS, EXAMPLE_PIECES)
+
+
+def test_pickle(toy_processor):
+    serialized = dumps(toy_processor)
+    deserialized = loads(serialized)
```
Comment (svlandeg) [inline curated_tokenizers/tests/test_bbpe_processor.py:122]: Perhaps we can also double check the correct class type of `deserialized` with an `assert isinstance`?
(same comment for the other 2 tests)

**Review by svlandeg — COMMENTED** [review]
Looks good! Just had a few small comments, feel free to merge otherwise.

**Comment by danieldk** [comment]
Note to self: needs to be backported to spaCy 3.x branch.

**Pull request merged**